# **Analisis Taxi Fare**

##### **Nama**  : Livia Amanda Annafiah

---

##### **Deskripsi Program**
##### Program ini bertujuan untuk mengambil informasi-informasi yang dibutuhkan dari database Taxi Fare dan melakukan pengolahan data.


# **ANSWER**

## **DATA LOADING**

In [157]:
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "hacktiv8-sql-417802"
client = bigquery.Client(project=project_id)

df = client.query('''
                  SELECT *
                  FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
                  ''').to_dataframe()

df

Authenticated


unique_key taxi_id      trip_start_timestamp        trip_end_timestamp  \
0        DP-250     D-7 2021-05-18 19:30:00+00:00 2021-05-18 19:45:00+00:00   
1        DP-289     D-7 2021-05-30 15:15:00+00:00 2021-05-30 15:45:00+00:00   
2        DP-316     D-7 2021-06-05 19:30:00+00:00 2021-06-05 19:45:00+00:00   
3        DP-354     D-7 2021-06-13 13:45:00+00:00 2021-06-13 14:15:00+00:00   
4        DP-501     D-7 2021-07-07 15:15:00+00:00 2021-07-07 15:30:00+00:00   
...         ...     ...                       ...                       ...   
9281    FI-2165   I-407 2021-04-06 16:00:00+00:00 2021-04-06 16:00:00+00:00   
9282    FI-3890   I-580 2022-11-13 15:30:00+00:00 2022-11-13 15:45:00+00:00   
9283   AI-11438  AI-527 2021-05-14 15:30:00+00:00 2021-05-14 15:30:00+00:00   
9284   AI-11220   A-842 2021-04-30 18:00:00+00:00 2021-04-30 18:00:00+00:00   
9285    FI-3315   I-580 2022-07-12 14:45:00+00:00 2022-07-12 14:45:00+00:00   

        trip_km   fare  tips payment_type        company  \
0      3.057746   9.00  0.00         Cash   Deluc Porter   
1     21.082354  35.50  0.00         Cash   Deluc Porter   
2      7.242030  14.00  0.00         Cash   Deluc Porter   
3     22.369826  37.25  0.00  Credit Card   Deluc Porter   
4      3.218680   9.00  0.00         Cash   Deluc Porter   
...         ...    ...   ...          ...            ...   
9281   1.287472   5.25  1.31  Credit Card  Federway Int.   
9282   2.574944   7.75  1.31  Credit Card  Federway Int.   
9283   1.448406   5.25  1.31  Credit Card    Atomic Inc.   
9284   1.770274   6.25  1.56  Credit Card    Atomic Inc.   
9285   1.609340   6.25  1.56  Credit Card  Federway Int.   

                           pickup_location  \
0      POINT (-87.6333080367 41.899602111)   
1      POINT (-87.913624596 41.9802643146)   
2      POINT (-87.913624596 41.9802643146)   
3      POINT (-87.913624596 41.9802643146)   
4      POINT (-87.6333080367 41.899602111)   
...                                    ...   
9281  POINT (-87.6209929134 41.8849871918)   
9282   POINT (-87.6173580061 41.859349715)   
9283  POINT (-87.6209929134 41.8849871918)   
9284  POINT (-87.6327464887 41.8809944707)   
9285  POINT (-87.6209929134 41.8849871918)   

                          dropoff_location  
0      POINT (-87.6635175498 41.874005383)  
1      POINT (-87.6950125892 42.001571027)  
2     POINT (-87.8045320063 41.9850151008)  
3     POINT (-87.6559981815 41.9442266014)  
4      POINT (-87.6635175498 41.874005383)  
...                                    ...  
9281  POINT (-87.6188683546 41.8909220259)  
9282  POINT (-87.6209929134 41.8849871918)  
9283  POINT (-87.6188683546 41.8909220259)  
9284   POINT (-87.657005027 41.8790669938)  
9285  POINT (-87.6378442095 41.8932163595)  

[9286 rows x 11 columns]

In [158]:
import pandas as pd

## **SQL**

##### 1. Apa saja nama perusahaan yang terdapat dalam data?

In [159]:
# Mencari nilai unik dari kolom company

df_1 = client.query('''
SELECT DISTINCT company as `Nama Perusahaan`
FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
''').to_dataframe()

df_1

Nama Perusahaan
0     Deluc Porter
1    Synctaxi Inc.
2      Atomic Inc.
3    Federway Int.
4  Rocket Cab Auto

Terdapat 5 perusahaan pada data Taxi Fare di atas, yaitu Deluc Porter, Synctaxi Inc., Atomic Inc., Federway Int., dan Rocket Cab Auto.

##### 2. Berapa nilai minimum dan maksimum dari waktu pengambilan data?

In [160]:
# Mencari nilai min dan max dari waktu pengambilan data

df_2 = client.query('''
SELECT MIN(trip_start_timestamp) as `Nilai Minimum`,
       MAX(trip_start_timestamp) as `Nilai Maximum`
FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
''').to_dataframe()

df_2

Nilai Minimum             Nilai Maximum
0 2021-01-02 11:00:00+00:00 2022-12-31 14:15:00+00:00

Nilai minimum dari seluruh perjalanan adalah pada tanggal 2 Januari 2021 pukul 11.00, sementara nilai maximum adalah pada tanggal 31 Desember 2022 pukul 14.15.

##### 3. Bandingkan jumlah banyaknya perjalanan antara seluruh perusahaan pada tahun 2022?

In [161]:
# Menampilkan banyaknya perjalanan berdasarkan perusahaan pada tahun 2022

df_3 = client.query('''
SELECT company as `Nama Perusahaan`,
       (COUNT(unique_key)) as `Jumlah Perjalanan`
FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
WHERE EXTRACT(YEAR FROM trip_start_timestamp) = 2022
GROUP BY 1
ORDER BY 2 DESC
''').to_dataframe()

df_3

Nama Perusahaan  Jumlah Perjalanan
0    Synctaxi Inc.               1341
1    Federway Int.               1059
2     Deluc Porter               1050
3  Rocket Cab Auto                 77

Berdasarkan data di atas, perusahaan yang memiliki banyaknya perjalanan terbanyak pada tahun 2022 adalah Synctaxi Inc. dengan total 1341 perjalanan, diikuti oleh Federway Int. dengan total  1059 perjalanan, diikuti oleh Deluc Porter dengan total 1050 perjalanan, dan terakhir adalah Rocket Cab Auto dengan total 77 perjalanan. Sementara itu, dari tabel di atas, terlihat bahwa perusahaan Deluc Porter tidak memiliki perjalanan pada tahun 2022.

##### 4. Perusahaan mana yang memiliki tarif perjalanan per km termurah?

In [162]:
# Menghitung tarif perjalanan per km termurah

df_4 = client.query('''
SELECT company as `Nama Perusahaan`,
       SUM(fare/trip_km) as `Tarif Perjalanan per KM`
FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
GROUP BY 1
ORDER BY 2 ASC
''').to_dataframe()

# Menghitung rata-rata tarif perjalanan per km termurah berdasarkan company

df_4b = client.query('''
SELECT company as `Nama Perusahaan`,
       AVG(fare/trip_km) as `Tarif Perjalanan per KM`
FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
GROUP BY 1
ORDER BY 2 ASC
''').to_dataframe()

# df_4 = client.query('''
# SELECT company as `Nama Perusahaan`,
#        IEEE_DIVIDE(fare,trip_km) as `Tarif Perjalanan per KM`
# FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
# WHERE 'Tarif Perjalanan per KM' != '0.0'
# ORDER BY 2 ASC
# ''').to_dataframe()

df_4
df_4b

Nama Perusahaan  Tarif Perjalanan per KM
0  Rocket Cab Auto                 2.488383
1     Deluc Porter                 3.831910
2      Atomic Inc.                 3.863726
3    Synctaxi Inc.                 3.945858
4    Federway Int.                 4.261964

Pada data di atas, tarif perjalanan per KM termurah dari seluruh perjalanan diraih oleh Rocket Cab Auto. Perusahaan selanjutnya yang memiliki rata-rata perjalanan per km termurah adalah Atomic Inc., diikuti oleh Synctaxi Inc., diikuti oleh Deluc Porter, dan terakhir adalah Federway Int.

Ketika menghitung rata-rata tarif per perjalanan dari masing-masing company untuk mengetahui company apa yang memiliki tarif termurah, didapatkan bahwa rata-rata termurah diraih oleh Rocket Cab Auto.


## **PANDAS**

In [163]:
# Data loading

df = client.query('''
              SELECT *
              FROM `ftds-hacktiv8-project.phase0_ftds_005_bsd.taxi_fare`
              LIMIT 3000
              ''').to_dataframe()

# Menampilkan dataframe
df

unique_key taxi_id      trip_start_timestamp        trip_end_timestamp  \
0        DP-250     D-7 2021-05-18 19:30:00+00:00 2021-05-18 19:45:00+00:00   
1        DP-289     D-7 2021-05-30 15:15:00+00:00 2021-05-30 15:45:00+00:00   
2        DP-316     D-7 2021-06-05 19:30:00+00:00 2021-06-05 19:45:00+00:00   
3        DP-354     D-7 2021-06-13 13:45:00+00:00 2021-06-13 14:15:00+00:00   
4        DP-501     D-7 2021-07-07 15:15:00+00:00 2021-07-07 15:30:00+00:00   
...         ...     ...                       ...                       ...   
2995    DP-2093   S-222 2022-10-01 20:15:00+00:00 2022-10-01 20:30:00+00:00   
2996    SI-1649   S-222 2022-11-01 10:15:00+00:00 2022-11-01 10:15:00+00:00   
2997     DP-134   S-228 2021-03-17 18:30:00+00:00 2021-03-17 18:30:00+00:00   
2998     DP-165   S-228 2021-03-29 18:15:00+00:00 2021-03-29 19:00:00+00:00   
2999     DP-264   S-228 2021-05-22 17:30:00+00:00 2021-05-22 18:30:00+00:00   

        trip_km   fare  tips payment_type        company  \
0      3.057746   9.00   0.0         Cash   Deluc Porter   
1     21.082354  35.50   0.0         Cash   Deluc Porter   
2      7.242030  14.00   0.0         Cash   Deluc Porter   
3     22.369826  37.25   0.0  Credit Card   Deluc Porter   
4      3.218680   9.00   0.0         Cash   Deluc Porter   
...         ...    ...   ...          ...            ...   
2995   1.448406   7.00   0.0         Cash   Deluc Porter   
2996   0.500000   3.25   0.0         Cash  Synctaxi Inc.   
2997   0.500000   4.25   0.0         Cash   Deluc Porter   
2998  27.358780  45.75   0.0  Credit Card   Deluc Porter   
2999  22.047958  37.75   0.0  Credit Card   Deluc Porter   

                           pickup_location  \
0      POINT (-87.6333080367 41.899602111)   
1      POINT (-87.913624596 41.9802643146)   
2      POINT (-87.913624596 41.9802643146)   
3      POINT (-87.913624596 41.9802643146)   
4      POINT (-87.6333080367 41.899602111)   
...                                    ...   
2995  POINT (-87.6262149064 41.8925077809)   
2996  POINT (-87.7381418906 41.9498764323)   
2997   POINT (-87.913624596 41.9802643146)   
2998  POINT (-87.9030396611 41.9790708201)   
2999  POINT (-87.9030396611 41.9790708201)   

                          dropoff_location  
0      POINT (-87.6635175498 41.874005383)  
1      POINT (-87.6950125892 42.001571027)  
2     POINT (-87.8045320063 41.9850151008)  
3     POINT (-87.6559981815 41.9442266014)  
4      POINT (-87.6635175498 41.874005383)  
...                                    ...  
2995  POINT (-87.6207628651 41.8983317935)  
2996  POINT (-87.7381418906 41.9498764323)  
2997   POINT (-87.913624596 41.9802643146)  
2998  POINT (-87.6209929134 41.8849871918)  
2999   POINT (-87.6603894557 41.958055933)  

[3000 rows x 11 columns]

##### 1. Buat kolom baru, dengan nama Keterangan_Waktu. Kolom ini berisikan label yang menunjukkan representasi keterangan waktu berdasarkan jam dengan ketentuan berikut:

In [164]:
# Mengecek nilai unik jam dari kolom trip_start_timestamp

df['trip_start_timestamp'].dt.hour.unique()

array([19, 15, 13, 20, 22, 18, 14, 16, 21,  9, 11, 17, 10, 12,  7, 23,  0,
        1,  8,  5,  2,  4,  6,  3])

In [165]:
# Menyimpan nilai kosong pada variabel
time = []

# Melakukan proses looping untuk menentukan kondisi
for i in df['trip_start_timestamp'].dt.hour:
  if i in (4, 5, 6, 7, 8, 9):
    time.append('Pagi')
  elif i in (10, 11, 12, 13, 14, 15):
    time.append('Siang')
  elif i in (16, 17, 18):
    time.append('Sore')
  elif i in (19, 20, 21, 22, 23):
    time.append('Malam')
  elif i in (0, 1, 2, 3):
    time.append('Dini Hari')

# Membuat kolom baru dengan variabel sebelumnya
df['Keterangan_Waktu'] = time

# Menampilkan dataframe terbaru
df

unique_key taxi_id      trip_start_timestamp        trip_end_timestamp  \
0        DP-250     D-7 2021-05-18 19:30:00+00:00 2021-05-18 19:45:00+00:00   
1        DP-289     D-7 2021-05-30 15:15:00+00:00 2021-05-30 15:45:00+00:00   
2        DP-316     D-7 2021-06-05 19:30:00+00:00 2021-06-05 19:45:00+00:00   
3        DP-354     D-7 2021-06-13 13:45:00+00:00 2021-06-13 14:15:00+00:00   
4        DP-501     D-7 2021-07-07 15:15:00+00:00 2021-07-07 15:30:00+00:00   
...         ...     ...                       ...                       ...   
2995    DP-2093   S-222 2022-10-01 20:15:00+00:00 2022-10-01 20:30:00+00:00   
2996    SI-1649   S-222 2022-11-01 10:15:00+00:00 2022-11-01 10:15:00+00:00   
2997     DP-134   S-228 2021-03-17 18:30:00+00:00 2021-03-17 18:30:00+00:00   
2998     DP-165   S-228 2021-03-29 18:15:00+00:00 2021-03-29 19:00:00+00:00   
2999     DP-264   S-228 2021-05-22 17:30:00+00:00 2021-05-22 18:30:00+00:00   

        trip_km   fare  tips payment_type        company  \
0      3.057746   9.00   0.0         Cash   Deluc Porter   
1     21.082354  35.50   0.0         Cash   Deluc Porter   
2      7.242030  14.00   0.0         Cash   Deluc Porter   
3     22.369826  37.25   0.0  Credit Card   Deluc Porter   
4      3.218680   9.00   0.0         Cash   Deluc Porter   
...         ...    ...   ...          ...            ...   
2995   1.448406   7.00   0.0         Cash   Deluc Porter   
2996   0.500000   3.25   0.0         Cash  Synctaxi Inc.   
2997   0.500000   4.25   0.0         Cash   Deluc Porter   
2998  27.358780  45.75   0.0  Credit Card   Deluc Porter   
2999  22.047958  37.75   0.0  Credit Card   Deluc Porter   

                           pickup_location  \
0      POINT (-87.6333080367 41.899602111)   
1      POINT (-87.913624596 41.9802643146)   
2      POINT (-87.913624596 41.9802643146)   
3      POINT (-87.913624596 41.9802643146)   
4      POINT (-87.6333080367 41.899602111)   
...                                    ...   
2995  POINT (-87.6262149064 41.8925077809)   
2996  POINT (-87.7381418906 41.9498764323)   
2997   POINT (-87.913624596 41.9802643146)   
2998  POINT (-87.9030396611 41.9790708201)   
2999  POINT (-87.9030396611 41.9790708201)   

                          dropoff_location Keterangan_Waktu  
0      POINT (-87.6635175498 41.874005383)            Malam  
1      POINT (-87.6950125892 42.001571027)            Siang  
2     POINT (-87.8045320063 41.9850151008)            Malam  
3     POINT (-87.6559981815 41.9442266014)            Siang  
4      POINT (-87.6635175498 41.874005383)            Siang  
...                                    ...              ...  
2995  POINT (-87.6207628651 41.8983317935)            Malam  
2996  POINT (-87.7381418906 41.9498764323)            Siang  
2997   POINT (-87.913624596 41.9802643146)             Sore  
2998  POINT (-87.6209929134 41.8849871918)             Sore  
2999   POINT (-87.6603894557 41.958055933)             Sore  

[3000 rows x 12 columns]

##### 2. Berapa total jarak perjalanan yang dicapai perusahaan Federway Int. pada tahun 2022? (Gunakan pandas query untuk menjawab soal no. 2)

In [166]:
# Menghitung total jarak perjalanan perusahaan

total = df[(df['company'] == 'Federway Int.') &
           (df['trip_start_timestamp'].dt.year == 2022)]['trip_km'].sum().round(2)

print(f"Total jarak perjalanan yang dicapai oleh perusahaan Federway Int. pada tahun 2022 adalah {total} km")

Total jarak perjalanan yang dicapai oleh perusahaan Federway Int. pada tahun 2022 adalah 1806.15 km


## **KESIMPULAN**

Kesimpulan yang dapat diambil dari keseluruhan proses ini adalah dataframe ini menggambarkan informasi data perjalanan dari taxi berdasarkan perusahaannya masing-masing. Terdapat 5 perusahaan pada data Taxi Fare di atas, yaitu Deluc Porter, Synctaxi Inc., Atomic Inc., Federway Int., dan Rocket Cab Auto. Perusahaan dengan tarif perjalanan per km termurah adalah Rocket Cab Auto. Perusahaan dengan jumlah perjalanan terbanyak pada tahun 2022 adalah Synctaxi Inc..